Improve login screen UI

james-hill-matomo opened the pull request:
### Description:

- Moved login form from twig to vue to enable auto-disable of submit.
- Auto disable submit.
- Much nicer design (Thanks Shelby!).

UI test failure fixed in https://github.com/matomo-org/tag-manager/pull/1009

PG-4133

### To test:

- Go to the login page when not logged in.
- See improved design.
- See login button disabled.
- Type something into user & pass fields.
- See login button enabled.
- Remove text from user or pass field.
- See login button disabled.

### Review

* [x] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [x] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [x] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [x] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [x] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [x] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [x] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [x] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

james-hill-matomo committed:
PG-4133 Improve login screen UI

- Moved form from twig to vue to enable auto-disable of submit.
- Auto disable submit.
- Much nicer design (Thanks Shelby!).

james-hill-matomo committed:
Updated tests for login form, PG-4133

james-hill-matomo commented:
@snake14 Can I get an initial review before we get core review?

- UI tests still to be updated.
- Some PHP tests failing unrelated to my PR, ignore those.

For some reason one of my CSS tweaks isn't showing up in screenshot tests - the Forgot Password link should be grey not blue - https://builds-artifacts.matomo.org/matomo-org/matomo/PG-4133-improve-ui-ux/14788482738/UIIntegrationTest_not_logged_in.png

Is this an artefact of our testing process?  It works on my local, and I can't see any ambiguities.  I can try tweaks to see if the CI system will pick up the change, but I thought it'd be best to get your input.

![image](https://github.com/user-attachments/assets/009d7174-b918-4281-bb34-3b3f716c452b)

![image](https://github.com/user-attachments/assets/b11a3340-9cb1-44f3-8f20-309232a79219)

snake14 commented on plugins/Login/vue/src/LoginForm/LoginForm.spec.ts:
Is there any way for this to work in the `tests/` directory instead of `vue/src/`?

james-hill-matomo replied:
I've deleted it for now as it's not part of our normal way of doing things, and the UI tests cover everything.

innocraft-automation committed:
Build vue files

james-hill-matomo committed:
Removed invalid spec.ts file, PG-4133

snake14 reviewed (commented):
@james-hill-matomo I didn't see any issues in my cursory review aside from the comment I left on the Vue spec file.

The 'Forgot password' hyperlink appears grey on my local instance. I couldn't get the UI test case to run locally due to a DDEV config issue.

james-hill-matomo committed:
UI test updates, PG-4133

james-hill-matomo committed:
UI test screenshots updated - PG-4133

james-hill-matomo committed:
Replaced commented out code, PG-4133

james-hill-matomo committed:
Merge branch '5.x-dev' into PG-4133-improve-ui-ux

james-hill-matomo committed:
Remove vue component and use plain js, PG-4133

james-hill-matomo committed:
Removed comment, PG-4133

james-hill-matomo committed:
Self-review fixes - PG-4133

james-hill-matomo committed:
Update screenshots + css tweak PG-4133

innocraft-automation committed:
Build vue files

AltamashShaikh committed:
Adds new whats new section entry for CustomReports, #PG-2528 (#23270)

james-hill-matomo committed:
UI test update, PG-4133

james-hill-matomo committed:
Merge branch '5.x-dev' into PG-4133-improve-ui-ux

james-hill-matomo committed:
Moved test to be local to correct plugin, PG-4133

james-hill-matomo committed:
Fix UI test result filename, PG-4133

sgiehl commented on plugins/TwoFactorAuth/tests/UI/expected-screenshots/TwoFactorAuth_logme_not_verified.png:
Shouldn't we for consistency reasons have the verify button also be auto-disabled and stretched to the full width? 🤔

james-hill-matomo replied:
@sgiehl Makes sense to me - can we get these changes in, and we can improve further in a separate ticket?

sgiehl replied:
We could do so, sure. But it actually happened too often that we ended up in a have baked state then, where the follow-up issue didn't get enough priority anymore to get resolved.

james-hill-matomo replied:
@sgiehl Ideally we'd have identified this up front and included the 2fa & reset password pages in the JIRA ticket.  My changes address all points in the JIRA ticket, but I see your point as to the completeness of the change.  You can see how far this work has drifted from the original intent already, which was simply to change the colour of the SAML button to more clearly differentiate it.

If we're to fully implement this, here is my plan:

- this ticket: Login (complete), forgot password (todo), 2fa (todo).
- scheduled for next sprint already: SAML login button (in plugin, not core repo)
- follow up "some day": set all buttons site wide to be bold, all disabled buttons to be darker shade of gray.

Is there anything else you consider a blocker before this PR can be merged?

PS - my view is that we should merge this before the SAML ticket (https://innocraft.atlassian.net/browse/PG-4276) is completed as it is already an improvement, but that's open for discussion.

Current:
![image](https://github.com/user-attachments/assets/a891e4dd-5d28-46d6-9a1c-d1fed08fd738)

With this PR: 
![image](https://github.com/user-attachments/assets/6fdfa020-88ad-4c8c-918f-bfdbab5a0132)

Final planned state:
![image](https://github.com/user-attachments/assets/dbce269a-d89c-4d39-a89e-6da244f737ea)

james-hill-matomo replied:
I've updated reset password & 2fa.  (and SSO signon in a separate PR)

sgiehl commented on plugins/Login/lang/en.json:
```diff
@@ -14,7 +14,9 @@
         "InvalidOrExpiredTokenV2": "The password has already been set or the link has expired.",
         "InvalidUsernameEmail": "Invalid username or e-mail address.",
         "LogIn": "Sign in",
+        "SignIn": "Sign In",
```
If the capital `I` is a fix and should be written that way everywhere, we should update the already existing translation above. For me there is no logical reason why that should be a capital `I`, though.

james-hill-matomo replied:
@sgiehl I was told not to update existing translations as doing so would not cause retranslation in other languages.  Maybe a moot point in this case, but I'd like to follow the process.

For all textual changes, I was following Shelby's design.  I'll loop her into these comments.

I also see your other comment that translations can be slow to come, and for an existing feature on a high traffic page, that seems like a very big deal!  What's the latest on solving that so that we can update translations without disruption to our users?  It seems like something we should fix.


Maybe it's best if I de-couple textual changes from this ticket, to be tackled elsewhere.

sgiehl replied:
Yes. We should not update existing translations, unless they are only fixing a typo. In that case retranslation for other languages won't be needed anyway.
I don't think Shelby is aware of the implications texual changes might cause, so good to take her in. 
But feel free to move all textual changes to a follow-up PR.

michalkleiner replied:
And in English, from a grammatical perspective, I don't think capitalising the 'I' makes sense, so even when there are things like this in designs, we need to double check those.

james-hill-matomo replied:
Maybe we need a new 2 stage process for text changes - first add the new translations to the en.json file, wait for translations to come back, then start using them.

Do we already do this?

james-hill-matomo replied:
I also hear that we don't have any SLA for translations to come back to us, so it could be any amount of time until translations are restored.

james-hill-matomo replied:
I'm leaving the text changes out of this PR.

sgiehl commented on plugins/Login/lang/en.json:
```diff
@@ -14,7 +14,9 @@
         "LogIn": "Sign in",
+        "SignIn": "Sign In",
         "LoginOrEmail": "Username or e-mail",
+        "EmailOrUsername": "Email or Username",
```
What's the clear benefit of writing `Email or Username` instead of `Username or e-mail`. We can for sure change that, but adding a new translation has the disadvantage that it will be untranslated for all languages afterwards - until someone adds translations for it. I personally see more disadvantage in loosing the translations than switching two words, without changing the meaning at all.

sgiehl commented on plugins/Login/lang/en.json:
```diff
@@ -36,6 +38,7 @@
         "CurrentlyBlockedIPsUnblockConfirm": "Unblock all blocked IPs?",
         "LoginPasswordNotCorrect": "Wrong username and/or password.",
         "LostYourPassword": "Lost your password?",
```
```suggestion
```
This one now seems unused an can be removed

sgiehl commented on plugins/Login/templates/login.twig:
```diff
@@ -21,47 +21,58 @@
                     <div class="row">
                         <div class="col s12 input-field">
-                            <input type="hidden" name="form_nonce" id="login_form_nonce" value="{{ nonce }}"/>
+                            <input type="hidden" name="form_nonce" id="login_form_nonce" value="{{  nonce }}"/>
```
```suggestion
                            <input type="hidden" name="form_nonce" id="login_form_nonce" value="{{ nonce }}"/>
```

sgiehl commented on plugins/Login/javascripts/login.js:
```diff
@@ -47,6 +47,14 @@
 
+        var disableOrEnableLoginSubmitButton = function () {
+          if ($('#login_form_login').val() === '' && $('#login_form_password').val() === '') {
+            $('#login_form_submit').addClass('disabled');
```
Is it useful to only add a class named `disabled`? Would it be more strict to add a disabled attribute instead/aswell? Or is there any reason why submitting the form should still work even with empty fields?

james-hill-matomo replied:
@sgiehl The class is used by the Materialize framework, and does indeed cause the button to be disabled.

I think it is much more correct and accessible to add the disabled attribute too, so I'll make the change.

sgiehl commented on plugins/Login/tests/UI/Login_spec.js:
```diff
@@ -67,6 +67,15 @@ describe("Login", function () {
 
+  // Enable login button
+  it('should enable the login button when username and password are entered', async function () {
+    await page.goto('?module=CoreHome&action=index&idSite=1&period=week&date=2017-06-04');
+    await page.type('#login_form_login', 'u');
+    await page.type('#login_form_password', 'p');
+
+    expect(await page.screenshot({ fullPage: true })).to.matchImage('login_button_enabled');
```
Assuming that this test actually should ensure that the login button gets enabled as soon as some input was provided, I personally would prefer this UI test not to take a screenshot.
Instead of taking a screenshot you could in the beginning of the test check that the button has the disabled attribute and in the end that it's gone.
That would also speed up the test a bit, as taking and comparing screenshots takes more time.

james-hill-matomo replied:
@sgiehl I agree

james-hill-matomo commented:
@sgiehl Thank you for the thorough review, I appreciate having this all up front so I can re-work without excessive back and forth.

james-hill-matomo committed:
Reset password button now matches login button, PG-4133

james-hill-matomo committed:
Updated to use disabled attribute, not css class

Also set the default to be disabled.

james-hill-matomo committed:
Merge branch '5.x-dev' into PG-4133-improve-ui-ux

james-hill-matomo committed:
Fixed spacing

james-hill-matomo committed:
Changed screenshot to functional test, PG-4133

james-hill-matomo committed:
Reverted text changes - moved to separate ticket, PG-4133

james-hill-matomo committed:
Updated test screenshots, PG-4133

james-hill-matomo committed:
Removed unnecessary UI screenshot file, PG-4133

james-hill-matomo committed:
Updated UI test screenshots, PG-4133

james-hill-matomo committed:
Darken disabled buttons, bold buttons on login. PG-4133

sgiehl commented on plugins/Login/stylesheets/login.less:
```diff
@@ -65,6 +65,14 @@
 
```
While testing locally I've seen that the login form currently always consumes 33% of the screen width. For a very very wide screen that imho now looks ugly, as the button stretches over the whole width.
I therefor would recommend to limit the max width of the form to a reasonable size, like e.g. 750px

```suggestion

    .contentForm {
        max-width: 750px;
    }

```

james-hill-matomo replied:
:+1:

sgiehl reviewed (changes requested):
The changes now look good to me. I've added a minor suggestion, which could be considered. Otherwise this looks good to merge.

james-hill-matomo committed:
2FA button stretched to match login buttons.

james-hill-matomo committed:
Minor less refactor, no functional change, PG-4133

james-hill-matomo committed:
Limit width of login screens to 750px max

james-hill-matomo committed:
Move button darkening into just the Login area for this PR

james-hill-matomo committed:
Updated UI tests, PG-4133

james-hill-matomo commented:
@sgiehl May I have a re-review please?

I've got a follow on PR to make sure we keep the disabled buttons consistent - https://github.com/matomo-org/matomo/pull/23305

I've split it away from this PR as it was a LOT of changed screenshots, and is out of scope of this ticket.


@spludlow For the disabled buttons, I've implemented the dark gray as "@color-gray" which is #ccc, not #cfcfcf .  Close enough IMO, and removes the complaint about inconsistent colouring.  I've set the disabled button to be this darker gray globally, but have only changed the buttons to use bold locally to the login page.

@spludlow I've ignored all textual changes as part of this ticket after Stefan's feedback.  The login wording is used in a few places, and will need some extra care.  Also there are translation challenges with changing existing wording.  Are you OK if we leave that for "some other time?".

![image](https://github.com/user-attachments/assets/a281f45b-d381-4cc4-8ec3-cb250941aa01)

james-hill-matomo committed:
Aligned login submit buttons with inputs above, PG-4133

james-hill-matomo commented:
@sgiehl I've made a tweak to the layout after feedback from @AltamashShaikh on the LoginSAML plugin changes, and have invalidated a bunch of my testing screenshots.  It's the end of day here, and I'm busy tomorrow, so these won't get fixed just yet.

sgiehl committed:
updates expected UI test files

sgiehl committed:
update submodule

sgiehl committed:
Merge branch '5.x-dev' into PG-4133-improve-ui-ux

sgiehl reviewed (approved):
I've updated the remaining UI tests. Overall this looks good to me now. We can do any further improvements that might be needed in follow-up PRs. I will merge this one now.
